src/datasets_misc/boston_housing.jl

```diff
@@ -41,77 +43,76 @@ Censoring is suggested by the fact that the highest median price of exactly \\\$
-
-julia> target[1]
-24.0
-```     
-"""
-function targets(; dir = nothing)
-    housing = readdlm(DATA, ',')
-    reshape(Vector{Float64}(housing[2:end,end]), (1, 506))
-end
-
-"""
-    feature_names()
-
-Return the  the names of the features provided in the dataset.
-"""
-function feature_names()
-    ["crim","zn","indus","chas","nox","rm","age","dis","rad","tax","ptratio","b","lstat"]
-end
-
-
-"""
-    features()
-
-Return the features of the Boston Housing dataset. This is a 13x506 Matrix of Float64 datatypes.
-The values are in the order ["crim","zn","indus","chas","nox","rm","age","dis","rad","tax","ptratio","b","lstat"].
-It has 506 examples.
 
+# Examples
+    
 ```jldoctest
 julia> using MLDatasets: BostonHousing
 
-julia> features = BostonHousing.features();
-
-julia> summary(features)
-"13×506 Matrix{Float64}"
+julia> dataset = BostonHousing()
+BostonHousing dataset:
+  feature_names : 13-element Vector{String}
+  features : 13×506 Matrix{Float64}
+  targets : 1×506 Matrix{Float64}
+
+julia> dataset.feature_names
+13-element Vector{String}:
+ "crim"
+ "zn"
+ "indus"
+ "chas"
+ "nox"
+ "rm"
+ "age"
+ "dis"
+ "rad"
+ "tax"
+ "ptratio"
+ "b"
+ "lstat"
+  
+julia> dataset[1:2]
+(features = [0.00632 0.02731; 18.0 0.0; … ; 396.9 396.9; 4.98 9.14], targets = [24.0 21.6])
+
+julia> X, y = dataset[];
 ```
 """
-function features()
-    housing = readdlm(DATA, ',')
-    Matrix{Float64}(housing[2:end,1:13])' |> collect
+struct BostonHousing <: AbstractDataset
+    _path::String
+    feature_names::Vector{String}
+    features::Matrix{Float64}
+    targets::Matrix{Float64}
 end
 
-end # module
+function BostonHousing()
+    path = joinpath(@__DIR__, "..", "..", "data", "boston_housing.csv")
+    feature_names = ["crim","zn","indus","chas","nox","rm","age","dis","rad","tax","ptratio","b","lstat"]
+    data = readdlm(path, ',')
+    features = Matrix{Float64}(data[2:end,1:13])' |> collect
+    targets = reshape(Vector{Float64}(data[2:end,end]), (1, 506))
+    return BostonHousing(path, feature_names, features, targets)
+end
 
+Base.getindex(d::BostonHousing) = getobs((; d.features, d.targets))
+Base.getindex(d::BostonHousing, i) = getobs((; d.features, d.targets), i)
+Base.length(d::BostonHousing) = numobs(d.features)
+
+# function Base.getproperty(::Type{BostonHousing}, s::Symbol)
+#     @show s
+#     d = BostonHousing()
+#     if s == :features
+#         Base.depwarn("BostonHousing.features() is deprecated, please use `BostonHousing().features` instead", :features)
+#         return d.features
+#     elseif s == :targets
+#         Base.depwarn("BostonHousing.targets() is deprecated, please use `BostonHousing().targets` instead", :targets)
+#         return d.targets
+#     elseif s == :feature_names
+#         Base.depwarn("BostonHousing.targets() is deprecated, please use `BostonHousing().feature_names` instead", :feature_names)
+#         return d.feature_names
+#     else 
+#         return getfield(d, s)
+#     end
+# end
```

Comment:
uncommenting in this, I get an error already at object creation, julia tries to access T.name and is rewired here.
```julia
julia> dataset = BostonHousing()
s = :name
Error showing value of type s = :name
ERROR: type BostonHousing has no field name
Stacktrace:
  [1] getproperty(#unused#::s = :name

SYSTEM (REPL): showing an error caused an error
ERROR: type BostonHousing has no field name
Stacktrace:
  [1] getproperty(#unused#::s = :name

SYSTEM (REPL): caught exception of type ErrorException while trying to handle a nested exception; giving up
```

Reply:
nevermind, there was a silly mistake